src/inventory_ui.cpp

```diff
@@ -40,409 +32,697 @@ struct itemstack_or_category {
     /** The item position in the players inventory. It should be unique as it
      * is used by the drop map in inventory_selector. */
     int item_pos;
-    itemstack_or_category(const indexed_invslice::value_type &a)
-        : it(&(a.first->front())), slice(a.first), category(&it->get_category()), item_pos(a.second)
-    {
+    size_t chosen_count = 0;
+
+    inventory_entry( const indexed_invslice::value_type &slice, const item_category *cat = nullptr )
+        : it( &( slice.first->front() ) ), slice( slice.first ),
+          category( ( cat != nullptr ) ? cat : &it->get_category() ), item_pos( slice.second ) {}
+    inventory_entry( const item *it, int pos, const item_category *cat = nullptr )
+        : it( it ), slice( nullptr ), category( ( cat != nullptr ) ? cat : &it->get_category() ), item_pos( pos ) {}
```

Comment:
Those defaults are unreadable. If you're going to have ternary operators inside them, it would be a good idea to put each of them on a separate line.